[Site Isolation][macOS] Propagate drag events to isolated frames

charliewolfe opened the pull request:
#### a1ad995cedf051ec5c280355b0a28696cdbba5d3
<pre>
[Site Isolation][macOS] Propagate drag events to isolated frames
<a href="https://bugs.webkit.org/show_bug.cgi?id=264502">https://bugs.webkit.org/show_bug.cgi?id=264502</a>
<a href="https://rdar.apple.com/118184993">rdar://118184993</a>

Reviewed by Wenson Hsieh.

This follows the same pattern as mouse events, where IPC is sent between the UI process and web
processes until the target frame is reached.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/page/DragClient.h:

* Source/WebCore/page/DragController.cpp:
(WebCore::DragController::dragExited):
(WebCore::DragController::dragEnteredOrUpdated):
(WebCore::DragController::tryDocumentDrag):
(WebCore::DragController::tryDHTMLDrag):
(WebCore::DragController::doSystemDrag):
(WebCore::DragController::dragEntered): Deleted.
(WebCore::DragController::dragUpdated): Deleted.
* Source/WebCore/page/DragController.h:

There were several places in `DragController` where drag functions would always be called on the main
frame’s event handler. This won’t work when the main frame is being hosted in another process. I
changed these functions to pass the frame which the drag is happening in, and instead call functions
on the event handler of that frame.

There were also places where we were down casting frames to local when we didn’t need to. In
`doSystemDrag()` and `tryDHTMLDrag()` all functions called on the main frame and main frame view are
accessible even if the frame is out-of-process.

`dragEntered()` and `dragUpdated()` didn’t seem like a useful abstraction, so I removed them.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::mouseEventDataForRemoteFrame):
(WebCore::EventHandler::dragSourceEndedAt):
* Source/WebCore/page/EventHandler.h:

I changed `dragSourceEndedAt()` to return a `FrameIdentifier` if the hit test leads to a remote frame.

* Source/WebCore/page/HandleMouseEventResult.h:
(WebCore::HandleMouseEventResult::HandleMouseEventResult):
(WebCore::HandleMouseEventResult::remoteInputEventData):
(WebCore::HandleMouseEventResult::remoteMouseEventData): Deleted.
* Source/WebCore/page/RemoteUserInputEventData.h: Renamed from Source/WebCore/page/RemoteMouseEventData.h.

I renamed `RemoteMouseEventData` to `RemoteUserInputEventData` since we can use the struct for things other
than mouse events. I used it here for drag events, and can likely be used for touch events too.

* Source/WebCore/platform/DragData.h:
(WebCore::DragData::setClientPosition):

I added `setClientPosition` to update the client positions coordinates on `DragData` when passing a drag
event to a remote frame.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WebPasteboardProxyCocoa.mm:
(WebKit::WebPasteboardProxy::determineDataOwner const):

When dragging in a isolated frame, each of the `MESSAGE_CHECK_COMPLETION(dataOwner, completionHandler())`
would fail. This is because `determineDataOwner()` is only checking for a `WebPageProxy` associated with
the `WebProcessProxy`, which may not exist. We also need to check each `RemotePageProxy`.

* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::sendMouseEvent):

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::performDragOperation):
(WebKit::WebPageProxy::performDragControllerAction):
(WebKit::WebPageProxy::dragEnded):
(WebKit::WebPageProxy::handleMouseEventReply):
(WebKit::WebPageProxy::sendMouseEvent):
* Source/WebKit/UIProcess/WebPageProxy.h:

The completion handler in `performDragControllerAction()` and `dragEnded()` is where the drag is
forwarded to the iframe process. I also added a frame identifier as a parameter so the same function
could be used to send the next drag.

`TestWebKitAPI.DragAndDropTests.DragAndDropOnEmptyView` expects drag events to be sent for empty webviews,
so we still need to send the drag IPC even if `m_mainFrame` is null. I made `FrameIdentifier` optional in
the drag IPC messages to the web page which will default to the main frame, since we don&apos;t know the frame
identifier of the main frame in the UI process in this case.

* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::remotePages const):
* Source/WebKit/UIProcess/WebProcessProxy.h:

I added `remotePages()` so they could be accessed in `WebPasteboardProxyCocoa`.

* Source/WebKit/WebProcess/WebCoreSupport/WebDragClient.cpp:
(WebKit::WebDragClient::startDrag):
* Source/WebKit/WebProcess/WebCoreSupport/WebDragClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/gtk/WebDragClientGtk.cpp:
(WebKit::WebDragClient::startDrag):
* Source/WebKit/WebProcess/WebCoreSupport/mac/WebDragClientMac.mm:
(WebKit::convertDragImageToBitmap):
(WebKit::WebDragClient::startDrag):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::mouseEvent):
(WebKit::WebPage::performDragControllerAction):
(WebKit::WebPage::dragEnded):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Change these functions to expect a frame identifier to be passed. If the frame identifer is null, use the
main frame.

* Source/WebKitLegacy/mac/WebCoreSupport/WebDragClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebDragClient.mm:
(WebDragClient::startDrag):
* Source/WebKitLegacy/mac/WebView/WebFrame.mm:
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _enteredDataInteraction:client:global:operation:]):
(-[WebView _updatedDataInteraction:client:global:operation:]):
(-[WebView _exitedDataInteraction:client:global:operation:]):
(-[WebView draggingEntered:]):
(-[WebView draggingUpdated:]):
(-[WebView draggingExited:]):

* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/270614@main">https://commits.webkit.org/270614@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b767fbb19bae788ead16b2e2a943d9b4cdc0b983

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25844 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4451 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27121 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27941 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23660 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26159 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6210 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1878 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23754 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26093 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3346 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22263 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28521 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2962 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23218 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29283 "layout-tests (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23579 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23594 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27156 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2989 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1212 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4380 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3445 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3325 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3305 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/Headers.cmake:
```diff
@@ -1534,7 +1534,7 @@ set(WebCore_PRIVATE_FRAMEWORK_HEADERS
     page/RemoteFrameClient.h
     page/RemoteFrameView.h
-    page/RemoteMouseEventData.h
+    page/RemoteInputEventData.h
```
Nit - `InputEvent` actually refers to a different type of editing-related DOM event in the specification, so this name might be a bit confusing. Since `WebCore::DragEvent` extends from `WebCore::MouseEvent`, it might be okay to just keep using `RemoteMouseEventData` for drag events; alternately, you could name it something like `RemoteUserInputEventData` or `RemoteMouseInputEventData`

charliewolfe replied:
This same struct will probably also be used for touch events in the future. I'll rename it to be `RemoteUserInputEventData`.

whsieh commented on Source/WebKitLegacy/mac/WebView/WebFrame.mm:
```diff
@@ -102,6 +102,7 @@
 #import <WebCore/PluginData.h>
 #import <WebCore/PrintContext.h>
 #import <WebCore/Range.h>
+#import <WebCore/RemoteInputEventData.h>
```
Is this import needed for unified sources?

charliewolfe replied:
It's needed for the `dragSourceEndedAt()` call in this file which now returns a `std::optional<WebCore::RemoteUserInputEventData>`.

whsieh commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -1302,6 +1302,59 @@ HTTPServer server({
+    [webView mouseDownAtPoint:beginPoint simulatePressure:NO];
+    [webView mouseMoveToPoint:endPoint withFlags:0];
+    [webView mouseUpAtPoint:endPoint];
+    [webView waitForPendingMouseEvents];
```
I wonder if it would be easier to use `DragAndDropSimulator` here (similar to the other tests in `DragAndDropTestsMac.mm`). It looks like you need a custom web view configuration here, but this constructor should do the trick:

```
- (instancetype)initWithWebViewFrame:(CGRect)frame configuration:(WKWebViewConfiguration *)configuration;
```

You can then do:

```
[simulator runFrom:CGPointMake(50, 50) to:CGPointMake(150, 150)];
```

…in case this doesn't Just Work™, I'd say adopting `DragAndDropSimulator` definitely doesn't block landing; however, it might mean that many of the other existing drag and drop tests could be failing with site isolation enabled, so we'll need to fix those eventually.

whsieh commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -1302,6 +1303,54 @@ HTTPServer server({
+    auto configuration = server.httpsProxyConfiguration();
+    enableSiteIsolation(configuration);
+    auto simulator = adoptNS([[DragAndDropSimulator alloc] initWithWebViewFrame:NSMakeRect(0, 0, 400, 400) configuration:configuration]);
+    auto webView = adoptNS([simulator webView]);
```
This is an overrelease, since `[simulator webView]` doesn't return a +1 object. We can just do this instead:
```suggestion
    RetainPtr webView = [simulator webView];
```

charliewolfe committed:
[Site Isolation][macOS] Propagate drag events to isolated frames
https://bugs.webkit.org/show_bug.cgi?id=264502
rdar://118184993

Reviewed by Wenson Hsieh.

This follows the same pattern as mouse events, where IPC is sent between the UI process and web
processes until the target frame is reached.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/page/DragClient.h:

* Source/WebCore/page/DragController.cpp:
(WebCore::DragController::dragExited):
(WebCore::DragController::dragEnteredOrUpdated):
(WebCore::DragController::tryDocumentDrag):
(WebCore::DragController::tryDHTMLDrag):
(WebCore::DragController::doSystemDrag):
(WebCore::DragController::dragEntered): Deleted.
(WebCore::DragController::dragUpdated): Deleted.
* Source/WebCore/page/DragController.h:

There were several places in `DragController` where drag functions would always be called on the main
frame’s event handler. This won’t work when the main frame is being hosted in another process. I
changed these functions to pass the frame which the drag is happening in, and instead call functions
on the event handler of that frame.

There were also places where we were down casting frames to local when we didn’t need to. In
`doSystemDrag()` and `tryDHTMLDrag()` all functions called on the main frame and main frame view are
accessible even if the frame is out-of-process.

`dragEntered()` and `dragUpdated()` didn’t seem like a useful abstraction, so I removed them.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::mouseEventDataForRemoteFrame):
(WebCore::EventHandler::dragSourceEndedAt):
* Source/WebCore/page/EventHandler.h:

I changed `dragSourceEndedAt()` to return a `FrameIdentifier` if the hit test leads to a remote frame.

* Source/WebCore/page/HandleMouseEventResult.h:
(WebCore::HandleMouseEventResult::HandleMouseEventResult):
(WebCore::HandleMouseEventResult::remoteInputEventData):
(WebCore::HandleMouseEventResult::remoteMouseEventData): Deleted.
* Source/WebCore/page/RemoteUserInputEventData.h: Renamed from Source/WebCore/page/RemoteMouseEventData.h.

I renamed `RemoteMouseEventData` to `RemoteUserInputEventData` since we can use the struct for things other
than mouse events. I used it here for drag events, and can likely be used for touch events too.

* Source/WebCore/platform/DragData.h:
(WebCore::DragData::setClientPosition):

I added `setClientPosition` to update the client positions coordinates on `DragData` when passing a drag
event to a remote frame.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WebPasteboardProxyCocoa.mm:
(WebKit::WebPasteboardProxy::determineDataOwner const):

When dragging in a isolated frame, each of the `MESSAGE_CHECK_COMPLETION(dataOwner, completionHandler())`
would fail. This is because `determineDataOwner()` is only checking for a `WebPageProxy` associated with
the `WebProcessProxy`, which may not exist. We also need to check each `RemotePageProxy`.

* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::sendMouseEvent):

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::performDragOperation):
(WebKit::WebPageProxy::performDragControllerAction):
(WebKit::WebPageProxy::dragEnded):
(WebKit::WebPageProxy::handleMouseEventReply):
(WebKit::WebPageProxy::sendMouseEvent):
* Source/WebKit/UIProcess/WebPageProxy.h:

The completion handler in `performDragControllerAction()` and `dragEnded()` is where the drag is
forwarded to the iframe process. I also added a frame identifier as a parameter so the same function
could be used to send the next drag.

`TestWebKitAPI.DragAndDropTests.DragAndDropOnEmptyView` expects drag events to be sent for empty webviews,
so we still need to send the drag IPC even if `m_mainFrame` is null. I made `FrameIdentifier` optional in
the drag IPC messages to the web page which will default to the main frame, since we don't know the frame
identifier of the main frame in the UI process in this case.

* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::remotePages const):
* Source/WebKit/UIProcess/WebProcessProxy.h:

I added `remotePages()` so they could be accessed in `WebPasteboardProxyCocoa`.

* Source/WebKit/WebProcess/WebCoreSupport/WebDragClient.cpp:
(WebKit::WebDragClient::startDrag):
* Source/WebKit/WebProcess/WebCoreSupport/WebDragClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/gtk/WebDragClientGtk.cpp:
(WebKit::WebDragClient::startDrag):
* Source/WebKit/WebProcess/WebCoreSupport/mac/WebDragClientMac.mm:
(WebKit::convertDragImageToBitmap):
(WebKit::WebDragClient::startDrag):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::mouseEvent):
(WebKit::WebPage::performDragControllerAction):
(WebKit::WebPage::dragEnded):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Change these functions to expect a frame identifier to be passed. If the frame identifer is null, use the
main frame.

* Source/WebKitLegacy/mac/WebCoreSupport/WebDragClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebDragClient.mm:
(WebDragClient::startDrag):
* Source/WebKitLegacy/mac/WebView/WebFrame.mm:
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _enteredDataInteraction:client:global:operation:]):
(-[WebView _updatedDataInteraction:client:global:operation:]):
(-[WebView _exitedDataInteraction:client:global:operation:]):
(-[WebView draggingEntered:]):
(-[WebView draggingUpdated:]):
(-[WebView draggingExited:]):

* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/270614@main
